src/pkcs11/openssl.c

```diff
@@ -405,7 +406,20 @@ CK_RV sc_pkcs11_verify_data(const unsigned char *pubkey, int pubkey_len,
 #endif
 	}
 
-	pkey = d2i_PublicKey(EVP_PKEY_RSA, NULL, &pubkey, pubkey_len);
+	/*
+	 * PKCS#11 does not define CKA_VALUE for public keys, and different cards
+	 * return either the direct or spki versions as defined in PKCS#15
+	 * And we need to support more then just RSA.
+	 * We can use d2i_PUBKEY which works for SPKI and any key type. 
+	 */
+	pubkey_tmp = pubkey; /* pass in so pubkey pointer is not modified */
+	pkey = d2i_PUBKEY(NULL, &pubkey_tmp, pubkey_len);
+
+	/* if failed, try old way that only works for RSA */
+	if  (pkey == NULL) {
+		pubkey_tmp = pubkey;
+		pkey = d2i_PublicKey(EVP_PKEY_RSA, NULL, &pubkey_tmp, pubkey_len);
```

Comment:
Could you check in OpenSSL whether using the "old method" is needed? I would expect that OpenSSL does this internally already...